src/Featurizers/TfidfVectorizerFeaturizer.h

```diff
@@ -0,0 +1,523 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include "Components/PipelineExecutionEstimatorImpl.h"
+#include "Components/DocumentStatisticsEstimator.h"
+#include "Structs.h"
+#include "../Traits.h"
+#include "../Strings.h"
+
+#include <cmath>
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         TfidfVectorizerTransformer
+///  \brief         Returns a unique TFIDFStruct for each input.
+///
+class TfidfVectorizerTransformer : public StandardTransformer<std::string, SparseVectorEncoding<std::float_t>> {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType                           = StandardTransformer<std::string, SparseVectorEncoding<std::float_t>>;
+    using IndexMap                           = std::unordered_map<std::string, std::uint32_t>;
+    using FrequencyMap                       = IndexMap;
+    using IterRangeType                      = std::tuple<std::string::const_iterator, std::string::const_iterator>;
+    using MapWithIterRange                   = std::map<IterRangeType, std::uint32_t, Components::IterRangeComp>;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    IndexMap const                           Labels;
+    FrequencyMap const                       DocumentFreq;
+    std::uint32_t const                      TotalNumsDocuments;
+    //Binary: If True, all non zero counts are set to 1. This is useful for discrete probabilistic models that model binary events rather than integer counts.
+    bool const                               Binary;
+    std::string const                        Norm;
+    bool const                               UseIdf;
+    bool const                               SmoothIdf;
+    bool const                               SublinearTf;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    TfidfVectorizerTransformer(
+        IndexMap labels,
+        FrequencyMap docufreq,
+        std::uint32_t totalnumdocus,
+        bool binary,
+        std::string norm,
```

Comment:
>norm [](start = 20, length = 4)

This should probable be an enum rather than a string. The compiler can provide checks for enums that can't be done with strings.

Reply:
Been an enum type, leave todo: for enum support in Traits. Will do after this PR